### ISSUE TITLE
Extending adding field to query

### ISSUE THREAD

**Issue by tobias-f** [graphql-java/graphql-java#1578]
Hello.

I am currently try to modify the schema generated from a SDL programatically by adding addition types, directives and so on. Without going too much into detail, I am able to get a reference to the generated schema (GraphQLSchema). Adding directives and types is no problem. However, I want to add also some additional queries to the Query type. I tried the following, but it fails with an exception:

    val schema = bean.schema
    val extendedQuery = schema.queryType.transform {
      it.field(... adding a field ...)
    }

    val extendedSchema = schema.transform {
       it.query(extendedQuery)
    }                                                        

`Caused by: graphql.AssertException: All types within a GraphQL schema must have unique names. No two provided types may have the same name.
No provided type may have a name which conflicts with any built in types (including Scalar and Introspection types).
You have redefined the type 'Query' from being a 'GraphQLObjectType' to a 'GraphQLObjectType'`

Any ideas how I can add additional queries to my schema programmatically?

Thanks a lot.

Cheers,

Tobias

**Comment by wasche** [comment]
I am running into this as well. I _think_ it might be due to GraphQLSchema.Builder.build() doing `schemaUtil.replaceTypeReferences(graphQLSchema);` (which looks to be directly altering the types) since we have a reference to the Query object via Mutation.query field. This is currently blocking me, so I will continue my investigation until I have a fix or workaround. My generated schema that I'm transforming is coming via SPQR, if that makes any difference.

**Comment by wasche** [comment]
I have found a work around for my issue, it seems to stem from how the schema was built through SPQR, as we allow things like:
```
mutation A {
  someMutation {
    field1
  }
  query {
    followOnQuery {
      fieldN
    }
  }
}
```

And it was the `query` field on the `Mutation` object that was causing issues when I was trying to rebuild the schema to add new query/mutation fields. I got around this by re-defining the query field on the Mutation object as a type reference.

**Comment by andimarek** [comment]
the general problem here is that the `GraphQLSchema` is an immutable graphq with possible cycles. When you want to change it you need to get back a whole new `GraphQLSchema` graph. This is currently not possible in an easy way, but we are planning to look into that more soon.

**Comment by andimarek** [comment]
The possibility for easy schema transformations will be available via #1605

**Comment by andimarek** [comment]
closing this issue now: please have a look at `SchemaTransformer` and `SchemaTransformerTest` for how to use the new feature. It will be available in 14.0
